File: modules/native-torch-transmitter/android/src/main/java/expo/modules/nativetorchtransmitter/NativeTorchTransmitterModule.kt
package expo.modules.nativetorchtransmitter

import android.content.Context
import android.hardware.camera2.CameraAccessException
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class NativeTorchTransmitterModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NativeTorchTransmitter")

    AsyncFunction("transmitBitstream") {
        bitstream: List<Int>, bitPeriodMs: Double, offsetMs: Double, promise: Promise ->

      val cameraManager = context.getSystemService(Context.CAMERA_SERVICE) as? CameraManager
      if (cameraManager == null) {
        promise.reject("E_NO_CAMERA", "CameraManager unavailable", null)
        return@AsyncFunction
      }

      // Find back camera with flash
      var cameraId: String? = null
      try {
        for (id in cameraManager.cameraIdList) {
          val chars = cameraManager.getCameraCharacteristics(id)
          val hasFlash = chars.get(CameraCharacteristics.FLASH_INFO_AVAILABLE) ?: false
          val isBack = chars.get(CameraCharacteristics.LENS_FACING) == CameraCharacteristics.LENS_FACING_BACK
          if (hasFlash && isBack) {
            cameraId = id
            break
          }
        }
      } catch (e: CameraAccessException) {
        promise.reject("E_CAMERA_ACCESS", "Failed to access camera: ${e.message}", e)
        return@AsyncFunction
      }

      if (cameraId == null) {
        promise.reject("E_NO_FLASH", "No back camera with flash found", null)
        return@AsyncFunction
      }

      val camId = cameraId
      val bits = bitstream.toIntArray()
      val periodNs = (bitPeriodMs * 1_000_000).toLong()
      val offsetNs = (offsetMs * 1_000_000).toLong()

      val thread = Thread {
        val periods = mutableListOf<Long>()
        try {
          // Pre-compute expected durations using the offset compensation algorithm
          val expectedDurations = OffsetCompensation.computeBitDurations(bits, periodNs, offsetNs)

          var nextDeadline = System.nanoTime() + expectedDurations[0]

          // Handle first bit
          if (bits.isNotEmpty()) {
            cameraManager.setTorchMode(camId, bits[0] == 1)
          }

          for (i in 1 until bits.size) {
            // Busy-wait until deadline
            while (System.nanoTime() < nextDeadline) {
              // spin
            }

            val actualTime = System.nanoTime()
            periods.add(actualTime - (nextDeadline - expectedDurations[i - 1]))

            cameraManager.setTorchMode(camId, bits[i] == 1)

            nextDeadline = actualTime + expectedDurations[i]
          }

          // Wait for last bit to complete
          while (System.nanoTime() < nextDeadline) {
            // spin
          }

          // Ensure torch is off
          cameraManager.setTorchMode(camId, false)

          // Calculate timing stats
          val periodMs = periods.map { it / 1_000_000.0 }
          val mean = if (periodMs.isNotEmpty()) periodMs.average() else 0.0
          val variance = if (periodMs.size > 1) {
            periodMs.map { (it - mean) * (it - mean) }.average()
          } else 0.0
          val stdDev = Math.sqrt(variance)
          val minP = periodMs.minOrNull() ?: 0.0
          val maxP = periodMs.maxOrNull() ?: 0.0

          val result = mapOf(
            "meanPeriodMs" to mean,
            "stdDevMs" to stdDev,
            "minPeriodMs" to minP,
            "maxPeriodMs" to maxP,
            "totalBits" to bits.size
          )

          promise.resolve(result)
        } catch (e: Exception) {
          try { cameraManager.setTorchMode(camId, false) } catch (_: Exception) {}
          promise.reject("E_TRANSMIT_FAILED", "Transmission failed: ${e.message}", e)
        }
      }

      thread.priority = Thread.MAX_PRIORITY
      thread.name = "WOP-TorchTransmitter"
      thread.start()
    }
  }

  private val context
    get() = requireNotNull(appContext.reactContext) { "Context is not available." }
}

File: modules/native-torch-transmitter/ios/NativeTorchTransmitterModule.swift
import ExpoModulesCore
import AVFoundation

public class NativeTorchTransmitterModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NativeTorchTransmitter")

    AsyncFunction("transmitBitstream") { (bitstream: [Int], bitPeriodMs: Double, offsetMs: Double, promise: Promise) in
      let status = AVCaptureDevice.authorizationStatus(for: .video)
      switch status {
      case .notDetermined:
        AVCaptureDevice.requestAccess(for: .video) { granted in
          if granted {
            self.doTransmit(bitstream: bitstream, bitPeriodMs: bitPeriodMs, offsetMs: offsetMs, promise: promise)
          } else {
            promise.reject("E_NO_PERMISSION", "Camera permission denied — required for torch access")
          }
        }
      case .authorized:
        self.doTransmit(bitstream: bitstream, bitPeriodMs: bitPeriodMs, offsetMs: offsetMs, promise: promise)
      case .denied:
        promise.reject("E_NO_PERMISSION", "Camera permission denied — enable in Settings > Privacy > Camera")
      case .restricted:
        promise.reject("E_NO_PERMISSION", "Camera access is restricted on this device")
      @unknown default:
        promise.reject("E_NO_PERMISSION", "Camera permission unavailable (unknown status)")
      }
    }
  }

  private func doTransmit(bitstream: [Int], bitPeriodMs: Double, offsetMs: Double, promise: Promise) {
    guard let device = AVCaptureDevice.default(for: .video) else {
      promise.reject("E_NO_DEVICE", "No video capture device found")
      return
    }
    guard device.hasTorch else {
      promise.reject("E_NO_FLASH", "This device does not have a torch")
      return
    }
    guard device.isTorchAvailable else {
      promise.reject("E_TORCH_UNAVAILABLE", "Torch is temporarily unavailable (device may be overheating)")
      return
    }

    let bits = bitstream
    let periodNs = UInt64(bitPeriodMs * 1_000_000)
    let offsetNs = UInt64(offsetMs * 1_000_000)

    let thread = Thread {
      var periods: [UInt64] = []
      var locked = false
      var torchError: Error? = nil

      do {
        try device.lockForConfiguration()
        locked = true

        func torchOn() throws {
          try device.setTorchModeOn(level: AVCaptureDevice.maxAvailableTorchLevel)
        }
        func torchOff() {
          device.torchMode = .off
        }

        var info = mach_timebase_info_data_t()
        mach_timebase_info(&info)
        let numer = UInt64(info.numer)
        let denom = UInt64(info.denom)

        func nanosToAbs(_ ns: UInt64) -> UInt64 {
          return ns * denom / numer
        }
        func absToNanos(_ abs: UInt64) -> UInt64 {
          return abs * numer / denom
        }

        let periodAbs = nanosToAbs(periodNs)
        let offsetAbs = nanosToAbs(offsetNs)

        // === Wake-up pulse: 200ms ON, 50ms OFF ===
        let wakeOnAbs = nanosToAbs(200_000_000)
        let wakeGapAbs = nanosToAbs(50_000_000)

        try torchOn()
        let wakeStart = mach_absolute_time()
        while mach_absolute_time() < wakeStart + wakeOnAbs { /* spin */ }

        torchOff()
        let gapStart = mach_absolute_time()
        while mach_absolute_time() < gapStart + wakeGapAbs { /* spin */ }

        // === Payload transmission ===
        // Pre-compute expected durations using the offset compensation algorithm
        let expectedDurations = OffsetCompensation.computeBitDurations(
          bits: bits, periodNs: periodNs, offsetNs: offsetNs
        ).map { nanosToAbs($0) }

        var nextDeadline = mach_absolute_time() + expectedDurations[0]

        if !bits.isEmpty {
          if bits[0] == 1 {
            try torchOn()
          } else {
            torchOff()
          }
        }

        for i in 1..<bits.count {
          while mach_absolute_time() < nextDeadline { /* spin */ }

          let actualTime = mach_absolute_time()
          periods.append(absToNanos(actualTime - (nextDeadline - expectedDurations[i - 1])))

          if bits[i] == 1 {
            try torchOn()
          } else {
            torchOff()
          }

          nextDeadline = actualTime + expectedDurations[i]
        }

        while mach_absolute_time() < nextDeadline { /* spin */ }

        torchOff()
        device.unlockForConfiguration()
        locked = false

        // Timing stats
        let periodMs = periods.map { Double($0) / 1_000_000.0 }
        let mean = periodMs.isEmpty ? 0.0 : periodMs.reduce(0, +) / Double(periodMs.count)
        let variance = periodMs.count > 1
          ? periodMs.map { ($0 - mean) * ($0 - mean) }.reduce(0, +) / Double(periodMs.count)
          : 0.0
        let stdDev = variance.squareRoot()
        let minP = periodMs.min() ?? 0.0
        let maxP = periodMs.max() ?? 0.0

        let result: [String: Any] = [
          "meanPeriodMs": mean,
          "stdDevMs": stdDev,
          "minPeriodMs": minP,
          "maxPeriodMs": maxP,
          "totalBits": bits.count,
        ]

        promise.resolve(result)
      } catch {
        if locked {
          device.torchMode = .off
          device.unlockForConfiguration()
        }
        promise.reject("E_TRANSMIT_FAILED", "Torch transmission failed: \(error.localizedDescription)")
      }
    }

    thread.qualityOfService = .userInteractive
    thread.name = "WOP-TorchTransmitter"
    thread.start()
  }
}
